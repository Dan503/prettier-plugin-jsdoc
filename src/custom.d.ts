declare namespace jest {
  interface Matchers<R, T> {
    toMatchSpecificSnapshot(filename: string): R;
  }
}
