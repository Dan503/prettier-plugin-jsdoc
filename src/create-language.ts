export default function createLanguage<T>(
  linguistData: T,
  override: (linguistData: T) => any,
): any {
  const { languageId, ...rest } = linguistData as any;
  return {
    linguistLanguageId: languageId,
    ...rest,
    ...override(linguistData),
  };
}
