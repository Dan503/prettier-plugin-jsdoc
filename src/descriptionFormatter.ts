import { format, BuiltInParserName } from "prettier";
import { DESCRIPTION, EXAMPLE, TODO } from "./tags";
import { AllOptions } from "./types";
import { capitalizer, formatCode } from "./utils";
import fromMarkdown from "mdast-util-from-markdown";
import { Root, Content } from "mdast";

const TABLE = "2@^5!~#sdE!_TABLE";

interface DescriptionEndLineParams {
  description: string;
  tag: string;
  isEndTag: boolean;
}

const parserSynonyms: Record<string, BuiltInParserName[]> = {
  js: ["babel", "babel-flow", "vue"],
  javascript: ["babel", "babel-flow", "vue"],
  ts: ["typescript", "babel-ts", "angular"],
  typescript: ["typescript", "babel-ts", "angular"],
  json: ["json", "json5", "json-stringify"],
};

function descriptionEndLine({
  description,
  tag,
  isEndTag,
}: DescriptionEndLineParams): string {
  if (description.trim().length < 0 || isEndTag) {
    return "";
  }

  if ([DESCRIPTION, EXAMPLE, TODO].includes(tag)) {
    return "\n";
  }

  return "";
}

interface FormatOptions {
  tagStringLength?: number;
}

/**
 * Trim, make single line with capitalized text. Insert dot if flag for it is
 * set to true and last character is a word character
 *
 * @private
 */
function formatDescription(
  tag: string,
  text: string,
  options: AllOptions,
  formatOptions: FormatOptions = {},
): string {
  if (!text) return text;

  const { printWidth, tsdoc } = options;
  const { tagStringLength = 0 } = formatOptions;

  // Wrap tag description
  const beginningSpace =
    tag === DESCRIPTION || (tag === EXAMPLE && tsdoc) ? "" : "  "; // google style guide space

  /**
   * change list with dash to dot for example:
   * 1- a thing
   *
   * to
   *
   * 1. a thing
   */
  text = text.replace(/^(\d+)[-][\s|]+/g, "$1. "); // Start
  text = text.replace(/\n+(\s*\d+)[-][\s]+/g, "\n$1. ");

  const tables: string[] = [];
  text = text.replace(/((\n|^)\|[\s\S]*?)((\n[^|])|$)/g, (code, _1, _2, _3) => {
    code = _3 ? code.slice(0, -1) : code;

    tables.push(code);
    return `\n\n${TABLE}\n\n${_3 ? _3.slice(1) : ""}`;
  });
  text = capitalizer(text);

  text = `${"!".repeat(tagStringLength)}${
    text.startsWith("```") ? "\n" : ""
  }${text}`;

  let tableIndex = 0;

  const rootAst = fromMarkdown(text);

  function stringyfy(
    mdAst: Content | Root,
    intention: string,
    parent: Content | Root | null,
  ): string {
    if (!Array.isArray(mdAst.children)) {
      if (mdAst.type === "inlineCode") {
        return ` \`${mdAst.value}\``;
      }

      if (mdAst.type === "code") {
        let result = mdAst.value || "";
        let _intention = intention;

        if (result) {
          // Remove two space from lines, maybe added previous format
          if (mdAst.lang) {
            const supportParsers = parserSynonyms[mdAst.lang.toLowerCase()];
            const parser = supportParsers?.includes(options.parser as any)
              ? options.parser
              : supportParsers?.[0] || mdAst.lang;

            result = formatCode(result, intention, {
              ...options,
              parser,
              jsdocKeepUnParseAbleExampleIndent: true,
            });
          } else {
            _intention = intention + " ".repeat(4);

            result = formatCode(result, _intention, {
              ...options,
              jsdocKeepUnParseAbleExampleIndent: true,
            });
          }
        }
        result = mdAst.lang ? result : result.trimEnd();
        return result
          ? mdAst.lang
            ? `\n\n${_intention}\`\`\`${mdAst.lang}${result}\`\`\``
            : `\n${result}`
          : "";
      }

      if (mdAst.value === TABLE) {
        if (parent) {
          parent.costumeType = TABLE;
        }

        if (tables.length > 0) {
          let result = tables?.[tableIndex] || "";
          tableIndex++;
          if (result) {
            result = format(result, {
              ...options,
              parser: "markdown",
            }).trim();
          }
          return `${
            result
              ? `\n\n${intention}${result.split("\n").join(`\n${intention}`)}`
              : mdAst.value
          }`;
        }
      }

      return (mdAst.value || "") as string;
    }
    return (mdAst.children as Content[])
      .map((ast, index) => {
        if (ast.type === "listItem") {
          let _listCount = `\n${intention}- `;
          // .replace(/((?!(^))\n)/g, "\n" + _intention);
          if (typeof mdAst.start === "number") {
            const count = index + ((mdAst.start as number) ?? 1);
            _listCount = `\n${intention}${count}. `;
          }

          const _intention = intention + " ".repeat(_listCount.length - 1);

          const result = stringyfy(ast, _intention, mdAst).trim();

          return `${_listCount}${result}`;
        }

        if (ast.type === "list") {
          return `\n${stringyfy(ast, intention, mdAst)}`;
        }

        if (ast.type === "paragraph") {
          let paragraph = stringyfy(ast, intention, parent);
          if (ast.costumeType === TABLE) {
            return paragraph;
          }
          paragraph = paragraph.replace(/\s+/g, " "); // Make single line

          paragraph = capitalizer(paragraph);
          if (options.jsdocDescriptionWithDot)
            paragraph = paragraph.replace(/([\w\p{L}])$/u, "$1."); // Insert dot if needed

          return `\n\n${breakDescriptionToLines(
            paragraph,
            printWidth,
            intention,
          )}`;
        }

        if (ast.type === "strong") {
          return `**${stringyfy(ast, intention, mdAst)}**`;
        }

        if (ast.type === "emphasis") {
          return `*${stringyfy(ast, intention, mdAst)}*`;
        }

        if (ast.type === "heading") {
          return `\n\n${intention}${"#".repeat(ast.depth)} ${stringyfy(
            ast,
            intention,
            mdAst,
          )}`;
        }

        if (ast.type === "link") {
          console.log({ link: JSON.stringify(ast) });
          return `[${stringyfy(ast, intention, mdAst)}](${ast.url})`;
        }

        return stringyfy(ast, intention, mdAst);
      })
      .join("");
  }

  let result = stringyfy(rootAst, beginningSpace, null);

  result = result.trim().slice(tagStringLength);

  return result;
}

function breakDescriptionToLines(
  desContent: string,
  maxWidth: number,
  beginningSpace: string,
): string {
  let str = desContent.trim();

  if (!str) {
    return str;
  }

  const extraLastLineWidth = 10;
  let result = "";
  while (str.length > maxWidth + extraLastLineWidth) {
    let sliceIndex = str.lastIndexOf(
      " ",
      str.startsWith("\n") ? maxWidth + 1 : maxWidth,
    );
    /**
     * When a str is a long word lastIndexOf will gives 4 every time loop
     * running unlimited time
     */
    if (sliceIndex <= beginningSpace.length)
      sliceIndex = str.indexOf(" ", beginningSpace.length + 1);

    if (sliceIndex === -1) sliceIndex = str.length;

    result += str.substring(0, sliceIndex);
    str = str.substring(sliceIndex + 1);

    str = `${beginningSpace}${str}`;
    str = `\n${str}`;
  }

  result += str;

  return `${beginningSpace}${result}`;
}

export { descriptionEndLine, FormatOptions, formatDescription };
