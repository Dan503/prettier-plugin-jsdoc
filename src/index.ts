import { getParser } from "./parser";
import parserBabel from "prettier/parser-babel";
import parserFlow from "prettier/parser-flow";
import parserTypescript from "prettier/parser-typescript";
import prettier from "prettier";

export const options = {
  jsdocParser: {
    type: "boolean",
    category: "jsdoc",
    default: true,
    description: "Format with jsdoc if is true",
  },
  jsdocSpaces: {
    type: "int",
    category: "jsdoc",
    default: 1,
    description: "How many spaces will be used to separate tag elements.",
  },
  jsdocDescriptionWithDot: {
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should dot be inserted at the end of description",
  },
  jsdocDescriptionTag: {
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should description tag be used",
  },
  jsdocVerticalAlignment: {
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should tags, types, names and description be aligned",
  },
  jsdocKeepUnParseAbleExampleIndent: {
    type: "boolean",
    category: "jsdoc",
    default: false,
    description:
      "Should unParseAble example (pseudo code or no js code) keep its indentation",
  },
};

const defaultOptions = {
  jsdocParser: true,
  jsdocSpaces: 1,
  jsdocDescriptionWithDot: false,
  jsdocDescriptionTag: false,
  jsdocVerticalAlignment: false,
  jsdocKeepUnParseAbleExampleIndent: false,
};

const languages = prettier
  .getSupportInfo()
  .languages.filter(({ name }) =>
    [
      "JavaScript",
      "Flow",
      "JSX",
      "TSX",
      "TypeScript",
      "Markdown",
      "MDX",
    ].includes(name),
  );

const parsers = {
  // JS - Babel
  get babel() {
    const parser = parserBabel.parsers.babel;
    return { ...parser, parse: getParser(parser.parse) };
  },
  get "babel-flow"() {
    const parser = parserBabel.parsers["babel-flow"];
    return { ...parser, parse: getParser(parser.parse) };
  },
  get "babel-ts"() {
    const parser = parserBabel.parsers["babel-ts"];
    return { ...parser, parse: getParser(parser.parse) };
  },
  // JS - Flow
  get flow() {
    const parser = parserFlow.parsers.flow;
    return { ...parser, parse: getParser(parser.parse) };
  },
  // JS - TypeScript
  get typescript() {
    const parser = parserTypescript.parsers.typescript;
    return { ...parser, parse: getParser(parser.parse) };
    // require("./parser-typescript").parsers.typescript;
  },
  get "jsdoc-parser"() {
    // Backward compatible, don't use this in new version since 1.0.0
    const parser = parserBabel.parsers["babel-ts"];
    return { ...parser, parse: getParser(parser.parse) };
  },
};

module.exports = {
  languages,
  options,
  parsers,
  defaultOptions,
};
