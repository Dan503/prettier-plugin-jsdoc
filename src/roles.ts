import {
  ABSTRACT,
  ASYNC,
  AUGMENTS,
  AUTHOR,
  BORROWS,
  CALLBACK,
  CATEGORY,
  CLASS,
  CONSTANT,
  DEFAULT,
  DEPRECATED,
  DESCRIPTION,
  EXAMPLE,
  EXTENDS,
  EXTERNAL,
  FILE,
  FIRES,
  FLOW,
  FUNCTION,
  IGNORE,
  LICENSE,
  MEMBER,
  MEMBEROF,
  MODULE,
  NAMESPACE,
  OVERRIDE,
  PARAM,
  PRIVATE,
  PRIVATE_REMARKS,
  PROPERTY,
  PROVIDES_MODULE,
  REMARKS,
  RETURNS,
  SEE,
  SINCE,
  TEMPLATE,
  THROWS,
  TODO,
  TYPE,
  TYPE_PARAM,
  TYPEDEF,
  VERSION,
  YIELDS,
} from "./tags";

const TAGS_SYNONYMS = {
  // One TAG TYPE can have different titles called SYNONYMS.  We want
  // to avoid different titles in the same tag so here is map with
  // synonyms as keys and tag type as value that we want to have in
  // final jsDoc.
  arg: PARAM,
  argument: PARAM,
  const: CONSTANT,
  constructor: CLASS,
  defaultvalue: DEFAULT,
  desc: DESCRIPTION,
  emits: FIRES,
  examples: EXAMPLE,
  exception: THROWS,
  fileoverview: FILE,
  func: FUNCTION,
  host: EXTERNAL,
  method: FUNCTION,
  overview: FILE,
  params: PARAM,
  prop: PROPERTY,
  return: RETURNS,
  var: MEMBER,
  virtual: ABSTRACT,
  yield: YIELDS,
  hidden: IGNORE,
};

const TAGS_NAMELESS = [
  BORROWS,
  CATEGORY,
  DEPRECATED,
  DESCRIPTION,
  EXAMPLE,
  EXTENDS,
  LICENSE,
  MODULE,
  NAMESPACE,
  OVERRIDE,
  RETURNS,
  SINCE,
  THROWS,
  TODO,
  YIELDS,
];

const TAGS_TYPELESS = [
  BORROWS,
  BORROWS,
  DEPRECATED,
  DESCRIPTION,
  EXAMPLE,
  IGNORE,
  LICENSE,
  MODULE,
  NAMESPACE,
  OVERRIDE,
  SINCE,
  TODO,
];

const TAGS_PEV_FORMATE_DESCRIPTION = [
  /** @todo should be formate like jsdoc standard saw https://jsdoc.app/tags-borrows.html  */
  BORROWS,
  DEFAULT,
  MEMBEROF,
  MODULE,
  SEE,
];

const TAGS_DESCRIPTION_NEEDED = [
  BORROWS,
  CATEGORY,
  DESCRIPTION,
  EXAMPLE,
  SINCE,
  TODO,
];

const TAGS_TYPE_NEEDED = [
  EXTENDS,
  PARAM,
  PROPERTY,
  RETURNS,
  THROWS,
  TYPE,
  TYPEDEF,
  YIELDS,
];

const TAGS_VERTICALLY_ALIGN_ABLE = [
  EXTENDS,
  PARAM,
  PROPERTY,
  RETURNS,
  THROWS,
  TYPE,
  TYPEDEF,
  YIELDS,
];

const TAGS_GROUP_HEAD = [CALLBACK, TYPEDEF];
const TAGS_GROUP_CONDITION = [
  ...TAGS_GROUP_HEAD,
  TYPE,
  PROPERTY,
  PARAM,
  RETURNS,
  YIELDS,
  THROWS,
];

const TAGS_ORDER = [
  PROVIDES_MODULE,
  MODULE,
  LICENSE,
  FLOW,
  ASYNC,
  PRIVATE,
  IGNORE,
  MEMBEROF,
  VERSION,
  AUTHOR,
  DEPRECATED,
  SINCE,
  CATEGORY,
  DESCRIPTION,
  REMARKS,
  PRIVATE_REMARKS,
  EXAMPLE,
  ABSTRACT,
  AUGMENTS,
  CONSTANT,
  DEFAULT,
  EXTERNAL,
  FILE,
  FIRES,
  TEMPLATE,
  TYPE_PARAM,
  FUNCTION,
  NAMESPACE,
  BORROWS,
  CLASS,
  EXTENDS,
  MEMBER,
  TYPEDEF,
  TYPE,
  PROPERTY,
  CALLBACK,
  PARAM,
  YIELDS,
  RETURNS,
  THROWS,
  "other",
  SEE,
  TODO,
];

export {
  TAGS_PEV_FORMATE_DESCRIPTION,
  TAGS_DESCRIPTION_NEEDED,
  TAGS_NAMELESS,
  TAGS_GROUP_HEAD,
  TAGS_GROUP_CONDITION,
  TAGS_ORDER,
  TAGS_SYNONYMS,
  TAGS_TYPE_NEEDED,
  TAGS_TYPELESS,
  TAGS_VERTICALLY_ALIGN_ABLE,
};
