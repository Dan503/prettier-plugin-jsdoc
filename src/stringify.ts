import { Tag } from "comment-parser";
import { format } from "prettier";
import {
  capitalizer,
  descriptionEndLine,
  EMPTY_LINE_SIGNATURE,
  NEW_LINE_START_THREE_SPACE_SIGNATURE,
  NEW_PARAGRAPH_START_THREE_SPACE_SIGNATURE,
} from "./utils";
import { DESCRIPTION, EXAMPLE, MEMBEROF, SEE } from "./tags";
import { TAGS_VERTICALLY_ALIGN_ABLE } from "./roles";
import { JsdocOptions, PrettierComment } from "./types";

const stringify = (
  { name, description, type, tag }: Tag,
  tagIndex: number,
  finalTagsArray: Tag[],
  options: JsdocOptions,
  comment: PrettierComment,
  maxTagTitleLength: number,
  maxTagTypeNameLength: number,
  maxTagNameLength: number,
): string => {
  const {
    loc: {
      start: { column },
    },
  } = comment;
  const gap = " ".repeat(options.jsdocSpaces);
  const { printWidth = 80 } = options;

  let tagTitleGapAdj = 0;
  let tagTypeGapAdj = 0;
  let tagNameGapAdj = 0;
  let descGapAdj = 0;

  if (
    options.jsdocVerticalAlignment &&
    TAGS_VERTICALLY_ALIGN_ABLE.includes(tag)
  ) {
    if (tag) tagTitleGapAdj += maxTagTitleLength - tag.length;
    else if (maxTagTitleLength) descGapAdj += maxTagTitleLength + gap.length;

    if (type) tagTypeGapAdj += maxTagTypeNameLength - type.length;
    else if (maxTagTypeNameLength)
      descGapAdj += maxTagTypeNameLength + gap.length;

    if (name) tagNameGapAdj += maxTagNameLength - name.length;
    else if (maxTagNameLength) descGapAdj = maxTagNameLength + gap.length;
  }

  const useTagTitle = tag !== DESCRIPTION || options.jsdocDescriptionTag;
  let tagString = "\n";

  if (useTagTitle) {
    tagString += `@${tag}${" ".repeat(tagTitleGapAdj || 0)}`;
  }
  if (type) {
    tagString += gap + `{${type}}` + " ".repeat(tagTypeGapAdj);
  }
  if (name) tagString += `${gap}${name}${" ".repeat(tagNameGapAdj)}`;

  // Add description (complicated because of text wrap)
  if (description && tag !== EXAMPLE) {
    if (useTagTitle) tagString += gap + " ".repeat(descGapAdj);
    if ([MEMBEROF, SEE].includes(tag)) {
      // Avoid wrapping
      tagString += description;
    } else {
      // Wrap tag description
      const beginningSpace = tag === DESCRIPTION ? "" : "    "; // google style guide space
      const marginLength = tagString.length;
      let maxWidth = printWidth - column - 3; // column is location of comment, 3 is ` * `

      if (marginLength >= maxWidth) {
        maxWidth = marginLength;
      }

      const resolveDescription = `${tagString}${description}`;

      tagString = resolveDescription
        .split(NEW_PARAGRAPH_START_THREE_SPACE_SIGNATURE)
        .map((newParagraph) => {
          return newParagraph
            .split(EMPTY_LINE_SIGNATURE)
            .map((paragraph) => {
              paragraph = capitalizer(paragraph);
              return paragraph
                .split(NEW_LINE_START_THREE_SPACE_SIGNATURE)
                .map((value) =>
                  breakDescriptionToLines(value, maxWidth, beginningSpace),
                )
                .join("\n    ");
            })
            .join("\n\n");
        })
        .join("\n\n    ");

      tagString = tagString ? `\n${tagString}` : tagString;
    }
  }

  // Try to use prettier on @example tag description
  if (tag === EXAMPLE) {
    const exampleCaption = description.match(
      /<caption>(((?!(<\/caption>))[\s\S])*)<\/caption>/i,
    );

    if (exampleCaption) {
      description = description.replace(exampleCaption[0], "");
      tagString = `${tagString} ${exampleCaption[0]}`;
    }

    try {
      let formattedExample = "";
      // if example is a json
      if (description.trim().startsWith("{")) {
        formattedExample = format(description || "", {
          ...options,
          parser: "json",
        });
      } else {
        formattedExample = format(description || "", options);
      }
      tagString += formattedExample.replace(/(^|\n)/g, "\n  ");
      tagString = tagString.slice(0, tagString.length - 3);
    } catch (err) {
      tagString += "\n";
      tagString += description
        .split("\n")
        .map(
          (l) =>
            `  ${options.jsdocKeepUnParseAbleExampleIndent ? l : l.trim()}`,
        )
        .join("\n");
    }
  }

  // Add empty line after some tags if there is something below
  tagString += descriptionEndLine({
    description: tagString,
    tag,
    isEndTag: tagIndex === finalTagsArray.length - 1,
  });

  return tagString;
};

export { stringify };

function breakDescriptionToLines(
  desContent: string,
  maxWidth: number,
  beginningSpace: string,
) {
  let str = desContent.trim();

  if (!str) {
    return str;
  }
  const extraLastLineWidth = 10;
  let result = "";
  while (str.length > maxWidth + extraLastLineWidth) {
    let sliceIndex = str.lastIndexOf(" ", maxWidth);
    /**
     * When a str is a long word lastIndexOf will gives 4 every time loop
     * running on limited time
     */
    if (sliceIndex <= beginningSpace.length)
      sliceIndex = str.indexOf(" ", beginningSpace.length + 1);

    if (sliceIndex === -1) sliceIndex = str.length;

    result += str.substring(0, sliceIndex);
    str = str.substring(sliceIndex + 1);

    str = `${beginningSpace}${str}`;
    str = `\n${str}`;
  }

  result += str;

  return result;
}
