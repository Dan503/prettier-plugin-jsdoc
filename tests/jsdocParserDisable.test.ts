import prettier from "prettier";
import { AllOptions } from "../src/types";

function subject(code: string, options: Partial<AllOptions> = {}) {
  return prettier.format(code, {
    plugins: [],
    parser: "babel",
    jsdocSpaces: 1,
    ...options,
  } as AllOptions);
}

test("template for callback", () => {
  const result = subject(`
/**
 * @template T
 *      @callback CallbackName
 *  @param {GetStyles<T>} getStyles
 * @returns       {UseStyle<T>}
 */
`);

  expect(result).toMatchSnapshot();
});

test("disabled complex object typedef ", () => {
  const result = subject(`
 /**
  * The bread crumbs indicate the navigate path and trigger the active page.
  * @class
  * @typedef {object} props
  * @prop        {any} navigation
  * @extends         {PureComponent<       props>} 
  */
 export class BreadCrumbs extends PureComponent {}
`);

  expect(result).toMatchSnapshot();

  const result2 = subject(
    `
  /**
   * The bread crumbs indicate the navigate path and trigger the active page.
   * @class
   * @typedef {object} props
   * @prop        {any} navigation
   * @extends         {PureComponent<       props>} 
   */
  export class BreadCrumbs extends PureComponent {}
 `,
    {
      plugins: ["prettier-plugin-jsdoc"],
      jsdocParser: false,
    },
  );

  expect(result).toBe(result2);
});
