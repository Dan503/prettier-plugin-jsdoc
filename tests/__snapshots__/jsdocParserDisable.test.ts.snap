// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extends 1`] = `
"/**
 * The bread crumbs indicate the navigate path and trigger the active page.
 * @class
 * @typedef {object} props
 * @prop        {any} navigation
 * @extends         {PureComponent<       props>}
 */
export class BreadCrumbs extends PureComponent {}
"
`;

exports[`template for callback 1`] = `
"/**
 * @template T
 *      @callback CallbackName
 *  @param {GetStyles<T>} getStyles
 * @returns       {UseStyle<T>}
 */
"
`;
